src/main/java/org/spongepowered/common/keyboard/KeyNetworkHandler.java

```diff
@@ -0,0 +1,168 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.keyboard;
+
+import com.google.common.collect.ImmutableList;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import net.minecraft.entity.player.EntityPlayerMP;
+import net.minecraft.network.NetHandlerPlayServer;
+import org.spongepowered.api.Platform;
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.entity.living.player.Player;
+import org.spongepowered.api.event.Listener;
+import org.spongepowered.api.event.SpongeEventFactory;
+import org.spongepowered.api.event.cause.Cause;
+import org.spongepowered.api.event.filter.cause.First;
+import org.spongepowered.api.event.keyboard.InteractKeyEvent;
+import org.spongepowered.api.event.keyboard.RegisterKeyBindingsEvent;
+import org.spongepowered.api.event.network.ChannelRegistrationEvent;
+import org.spongepowered.api.keyboard.KeyBinding;
+import org.spongepowered.api.keyboard.KeyContexts;
+import org.spongepowered.api.network.PlayerConnection;
+import org.spongepowered.api.network.RemoteConnection;
+import org.spongepowered.common.SpongeImpl;
+import org.spongepowered.common.interfaces.entity.player.IMixinEntityPlayerMP;
+import org.spongepowered.common.network.message.MessageGuiState;
+import org.spongepowered.common.network.message.MessageKeyState;
+import org.spongepowered.common.network.message.MessageKeyboardData;
+import org.spongepowered.common.network.message.SpongeMessageHandler;
+import org.spongepowered.common.registry.type.keyboard.KeyBindingRegistryModule;
+
+import java.util.BitSet;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class KeyNetworkHandler {
+
+    @Listener
+    public void onRegisterChannel(ChannelRegistrationEvent.Register event, @First PlayerConnection connection) {
+        if (event.getChannel().equals(SpongeMessageHandler.CHANNEL_NAME)) {
+            handleKeyRegistration(connection);
+        }
+    }
+
+    public static void handleKeyRegistration(PlayerConnection connection) {
+        final Player player = connection.getPlayer();
+        // TODO: Why is the connection still null at this point, is this normal in sp (SpongeForge) ???
```

Comment:
Well this is rather annoying because you cannot send packets directly to a `PlayerConnection`, you have to pass the player instance through in the `ChannelBinding`, meaning in the SpongeAPI, you cannot use the event to send packets to the client. IMO the player connection should be set for the player during this event, even if it's reset afterwards (like I did here to work around the issue).